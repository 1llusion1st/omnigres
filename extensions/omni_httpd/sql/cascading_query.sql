create table routes
(
    name     text not null,
    query    text not null,
    priority int  not null
);

insert
into
    routes (name, query, priority)
values
    ('test', $$SELECT omni_httpd.http_response(body => 'test') FROM request WHERE request.path = '/test'$$, 1),
    ('ping', $$SELECT omni_httpd.http_response(body => 'pong') FROM request WHERE request.path = '/ping'$$, 1);

\pset format wrapped
\pset columns 80

-- Preview the query
select
    omni_httpd.cascading_query(name, query order by priority desc nulls last)
from
    routes;

\pset format aligned

-- Try it
begin;
with
    listener as (insert into omni_httpd.listeners (address, port) values ('127.0.0.1', 9100) returning id),
    handler as (insert into omni_httpd.handlers (query) select
                                                            omni_httpd.cascading_query(name, query
                                                                                       order by priority desc nulls last)
                                                        from
                                                            routes returning id)
insert
into
    omni_httpd.listeners_handlers (listener_id, handler_id)
select
    listener.id,
    handler.id
from
    listener,
    handler;
delete
from
    omni_httpd.configuration_reloads;
end;

call omni_httpd.wait_for_configuration_reloads(1);

\! curl --retry-connrefused --retry 10  --retry-max-time 10 -w '\n\n' --silent http://localhost:9100/test

\! curl --retry-connrefused --retry 10  --retry-max-time 10 -w '\n\n' --silent http://localhost:9100/ping

-- CTE handling

\pset format wrapped
\pset columns 80

select
    omni_httpd.cascading_query(name, query order by priority desc nulls last)
from
    (values
         ('test',
          $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'test') FROM request, Test WHERE request.path = '/test' and test.val = 1$$,
          1),
         ('ping',
          $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'pong') FROM request, Test WHERE request.path = '/ping' and test.val = 1$$,
          1))
        as routes(name, query, priority);

\pset format aligned