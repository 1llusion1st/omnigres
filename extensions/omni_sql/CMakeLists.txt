cmake_minimum_required(VERSION 3.25.1)
project(omni_sql)

if(NOT DEFINED ENV{OMNI_SQL_ALREADY_INCLUDED})
    set(ENV{OMNI_SQL_ALREADY_INCLUDED} TRUE)

    list(PREPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/../../cmake)

    include(CPM)
    include(CTest)

    enable_testing()

    find_package(PostgreSQL REQUIRED)

    add_library(libomnisql STATIC
            deparse.c deparse_15.c deparse_14.c deparse_13.c
            lib.c)
    set_property(TARGET libomnisql PROPERTY POSITION_INDEPENDENT_CODE ON)
    target_include_directories(libomnisql
            PRIVATE ${PostgreSQL_SERVER_INCLUDE_DIRS}
            PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})

    add_postgresql_extension(
            omni_sql
            VERSION 0.1
            SCHEMA omni_sql
            RELOCATABLE false
            SCRIPTS omni_sql--0.1.sql
            SOURCES omni_sql.c
            DEPENDS_ON libpgaug
            REGRESS deparse cte parameterized validity
    )

    target_link_libraries(omni_sql libpgaug libomnisql)

endif()