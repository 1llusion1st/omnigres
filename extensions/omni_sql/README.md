# omni_sql

SQL statement processing