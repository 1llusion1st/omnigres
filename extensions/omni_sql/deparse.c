// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on

#include "lib/stringinfo.h"
#include "nodes/nodes.h"
#include "nodes/parsenodes.h"

void omni_sql_deparseRawStmt(StringInfo str, RawStmt *raw_stmt);

char *omni_sql_deparse_statement(List *stmts) {
  char *result;
  StringInfoData str;
  ListCell *lc;

  initStringInfo(&str);

  foreach (lc, stmts) {
    omni_sql_deparseRawStmt(&str, castNode(RawStmt, lfirst(lc)));
    if (lnext(stmts, lc))
      appendStringInfoString(&str, "; ");
  }
  result = strdup(str.data);

  return result;
}
