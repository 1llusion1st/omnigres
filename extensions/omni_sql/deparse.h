#ifndef OMNI_SQL_DEPARSE_H
#define OMNI_SQL_DEPARSE_H

// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on

char *omni_sql_deparse_statement(List *stmts);
#endif // OMNI_SQL_DEPARSE_H