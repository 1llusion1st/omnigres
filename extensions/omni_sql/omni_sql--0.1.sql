CREATE TYPE statement;

CREATE FUNCTION statement_in(cstring) RETURNS statement
  AS 'MODULE_PATHNAME', 'statement_in'
  LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION statement_out(statement) RETURNS cstring
  AS 'MODULE_PATHNAME', 'statement_out'
  LANGUAGE C STRICT IMMUTABLE;

CREATE TYPE statement (
    INPUT = statement_in,
    OUTPUT = statement_out,
    LIKE = text
);

CREATE FUNCTION add_cte(statement, name text, cte statement,
                        recursive bool DEFAULT false, prepend bool DEFAULT false) RETURNS statement
  AS 'MODULE_PATHNAME', 'add_cte'
  LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION is_parameterized(statement) RETURNS bool
  AS 'MODULE_PATHNAME', 'is_parameterized'
  LANGUAGE C STRICT IMMUTABLE;