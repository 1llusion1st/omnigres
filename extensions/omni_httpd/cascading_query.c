// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on

#include <omni_sql.h>

PG_FUNCTION_INFO_V1(cascading_query_reduce);

PG_FUNCTION_INFO_V1(cascading_query_final);

#define ARG_STATE 0
#define ARG_NAME 1
#define ARG_QUERY 2

Datum cascading_query_reduce(PG_FUNCTION_ARGS) {
  List *stmts;

  MemoryContext memory_context;
  if (AggCheckCallContext(fcinfo, &memory_context) != 1) {
    ereport(ERROR, errmsg("must be called as a regular aggregate"));
  }

  MemoryContext old_context = MemoryContextSwitchTo(memory_context);

  if PG_ARGISNULL (ARG_NAME) {
    ereport(ERROR, errmsg("name can't be NULL"));
  }

  if PG_ARGISNULL (ARG_QUERY) {
    ereport(ERROR, errmsg("query can't be NULL"));
  }

  // We already have the query
  bool initialized = true;

  if PG_ARGISNULL (ARG_STATE) {
    stmts = omni_sql_parse_statement("SELECT");
    initialized = false;
  } else {
    stmts = (List *)PG_GETARG_POINTER(ARG_STATE);
  }

  text *name = PG_GETARG_TEXT_PP(ARG_NAME);
  char *query = text_to_cstring(PG_GETARG_TEXT_PP(ARG_QUERY));

  RawStmt *original_stmt = castNode(RawStmt, linitial(stmts));
  SelectStmt *original_select_stmt = castNode(SelectStmt, original_stmt->stmt);

  // SELECT * FROM {name}
  SelectStmt *select = makeNode(SelectStmt);
  ResTarget *resTarget = makeNode(ResTarget);
  ColumnRef *star = makeNode(ColumnRef);
  star->fields = list_make1(makeNode(A_Star));
  resTarget->val = (Node *)star;
  select->targetList = list_make1(resTarget);
  RangeVar *from = makeNode(RangeVar);
  from->inh = true;
  from->relname = text_to_cstring(name);
  select->fromClause = list_make1(from);

  if (!initialized) {
    original_stmt->stmt = (Node *)select;
  } else {
    char *previous_name =
        llast_node(CommonTableExpr, original_select_stmt->withClause->ctes)->ctename;

    // SELECT true FROM {previous_name}
    SelectStmt *subQuery = makeNode(SelectStmt);
    ResTarget *subResTarget = makeNode(ResTarget);
    ColumnRef *subRes = makeNode(ColumnRef);
    subResTarget->val = (Node *)star;
    RangeVar *subFrom = makeNode(RangeVar);
    subFrom->inh = true;
    subFrom->relname = previous_name;
    subQuery->fromClause = list_make1(subFrom);

    // NOT EXISTS (SELECT FROM {previous_name})
    BoolExpr *parent = makeNode(BoolExpr);
    parent->boolop = NOT_EXPR;
    SubLink *subLink = makeNode(SubLink);
    subLink->subLinkType = EXISTS_SUBLINK;
    subLink->subselect = (Node *)subQuery;
    parent->args = list_make1(subLink);
    select->whereClause = (Node *)parent;

    // {existing_query} UNION ALL SELECT * FROM {name} WHERE NOT EXISTS (SELECT FROM {previous_name}
    SelectStmt *setop = castNode(SelectStmt, makeNode(SelectStmt));

    setop->op = SETOP_UNION;
    setop->larg = original_select_stmt;
    setop->rarg = select;
    setop->withClause = original_select_stmt->withClause;
    original_select_stmt->withClause = NULL;

    original_stmt->stmt = (Node *)setop;
  }

  omni_sql_add_cte(stmts, name, omni_sql_parse_statement(query), false, false);

  MemoryContextSwitchTo(old_context);

  PG_RETURN_POINTER(stmts);
}

Datum cascading_query_final(PG_FUNCTION_ARGS) {
  List *stmts;

  if PG_ARGISNULL (ARG_STATE) {
    stmts = omni_sql_parse_statement("SELECT");
  } else {
    stmts = (List *)PG_GETARG_POINTER(ARG_STATE);
  }
  PG_RETURN_TEXT_P(cstring_to_text(omni_sql_deparse_statement(stmts)));
}
