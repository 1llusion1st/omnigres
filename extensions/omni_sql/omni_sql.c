/**
 * @file omni_sql.c
 *
 */

// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on
#include <miscadmin.h>

#include "omni_sql.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(statement_in);

Datum statement_in(PG_FUNCTION_ARGS) {
  if (PG_ARGISNULL(0)) {
    ereport(ERROR, errmsg("statement can't be NULL"));
  }
  char *statement = PG_GETARG_CSTRING(0);
  List *stmts = omni_sql_parse_statement(statement);
  char *deparsed = omni_sql_deparse_statement(stmts);
  text *deparsed_statement = cstring_to_text(deparsed);
  PG_RETURN_TEXT_P(deparsed_statement);
}

PG_FUNCTION_INFO_V1(statement_out);

Datum statement_out(PG_FUNCTION_ARGS) { PG_RETURN_CSTRING(text_to_cstring(PG_GETARG_TEXT_PP(0))); }

PG_FUNCTION_INFO_V1(add_cte);

Datum add_cte(PG_FUNCTION_ARGS) {
  if (PG_ARGISNULL(0)) {
    ereport(ERROR, errmsg("Statement should not be NULL"));
  }
  if (PG_ARGISNULL(1)) {
    ereport(ERROR, errmsg("CTE name should not be NULL"));
  }
  if (PG_ARGISNULL(2)) {
    ereport(ERROR, errmsg("CTE should not be NULL"));
  }
  if (PG_ARGISNULL(3)) {
    ereport(ERROR, errmsg("Recursive flag should not be NULL"));
  }

  if (PG_ARGISNULL(4)) {
    ereport(ERROR, errmsg("Prepend flag should not be NULL"));
  }

  text *statement = PG_GETARG_TEXT_PP(0);
  char *cstatement = text_to_cstring(statement);
  List *stmts = omni_sql_parse_statement(cstatement);

  text *cte_name = PG_GETARG_TEXT_PP(1);

  text *cte = PG_GETARG_TEXT_PP(2);
  char *ccte = text_to_cstring(cte);
  List *cte_stmts = omni_sql_parse_statement(ccte);

  bool recursive = PG_GETARG_BOOL(3);
  bool prepend = PG_GETARG_BOOL(4);

  stmts = omni_sql_add_cte(stmts, text_to_cstring(cte_name), cte_stmts, recursive, prepend);

  char *deparsed = omni_sql_deparse_statement(stmts);
  text *deparsed_statement = cstring_to_text(deparsed);

  PG_RETURN_TEXT_P(deparsed_statement);
}

PG_FUNCTION_INFO_V1(is_parameterized);

Datum is_parameterized(PG_FUNCTION_ARGS) {
  if (PG_ARGISNULL(0)) {
    ereport(ERROR, errmsg("statement can't be NULL"));
  }
  text *statement = PG_GETARG_TEXT_PP(0);
  List *stmts = omni_sql_parse_statement(text_to_cstring(statement));

  PG_RETURN_BOOL(omni_sql_is_parameterized(stmts));
}

PG_FUNCTION_INFO_V1(is_valid);

Datum is_valid(PG_FUNCTION_ARGS) {
  if (PG_ARGISNULL(0)) {
    ereport(ERROR, errmsg("statement can't be NULL"));
  }
  text *statement = PG_GETARG_TEXT_PP(0);
  char *string = text_to_cstring(statement);
  List *stmts = omni_sql_parse_statement(string);

  PG_RETURN_BOOL(omni_sql_is_valid(stmts, NULL));
}

PG_FUNCTION_INFO_V1(raw_statements);

static inline void find_line_col(char *str, int offset, int *line, int *col) {
  *line = 1;
  *col = 1;

  for (int i = 0; i < offset; i++) {
    if (str[i] == '\0') {
      return;
    }
    if (str[i] == '\n') {
      (*line)++;
      *col = 1;
    } else {
      (*col)++;
    }
  }
}

static inline int find_non_whitespace(char *str) {
  int offset = 0;

  while (str[offset]) {
    if (!isspace((unsigned char)str[offset])) {
      return offset;
    }
    offset++;
  }

  // If only whitespace, use it as is (but this shouldn't happen, there's no statement!)
  return 0;
}

Datum raw_statements(PG_FUNCTION_ARGS) {
  if (PG_ARGISNULL(0)) {
    ereport(ERROR, errmsg("statements can't be NULL"));
  }

  ReturnSetInfo *rsinfo = (ReturnSetInfo *)fcinfo->resultinfo;
  rsinfo->returnMode = SFRM_Materialize;

  MemoryContext per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
  MemoryContext oldcontext = MemoryContextSwitchTo(per_query_ctx);

  Tuplestorestate *tupstore = tuplestore_begin_heap(false, false, work_mem);
  rsinfo->setResult = tupstore;

  char *statement = PG_GETARG_CSTRING(0);

  List *stmts = omni_sql_parse_statement(statement);

  ListCell *lc;
  foreach (lc, stmts) {
    switch (nodeTag(lfirst(lc))) {
    case T_RawStmt: {
      RawStmt *raw_stmt = lfirst_node(RawStmt, lc);
      int line, col;
      int actual_start = find_non_whitespace(statement + raw_stmt->stmt_location);
      find_line_col(statement, raw_stmt->stmt_location + actual_start, &line, &col);
      Datum values[3] = {
          PointerGetDatum(
              raw_stmt->stmt_len == 0
                  ? cstring_to_text(statement + raw_stmt->stmt_location + actual_start)
                  : cstring_to_text_with_len(statement + raw_stmt->stmt_location + actual_start,
                                             raw_stmt->stmt_len)),
          Int32GetDatum(line), Int32GetDatum(col)};
      bool isnull[3] = {false, false, false};
      tuplestore_putvalues(tupstore, rsinfo->expectedDesc, values, isnull);
      break;
    }
    default:
      break;
    }
  };

  tuplestore_donestoring(tupstore);

  MemoryContextSwitchTo(oldcontext);
  PG_RETURN_NULL();
}
