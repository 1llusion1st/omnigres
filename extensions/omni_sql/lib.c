/**
 * @file lib.c
 *
 */

// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on

#include <funcapi.h>
#include <parser/parser.h>
#include <utils/builtins.h>

#include <nodes/nodeFuncs.h>

#include "deparse.h"

List *omni_sql_parse_statement(char *statement) {
#if PG_MAJORVERSION_NUM == 13
  List *stmts = raw_parser(statement);
#else
  List *stmts = raw_parser(statement, RAW_PARSE_DEFAULT);
#endif
  return stmts;
}

List *omni_sql_add_cte(List *stmts, text *cte_name, List *cte_stmts, bool recursive, bool prepend) {

  if (list_length(stmts) != 1) {
    ereport(ERROR, errmsg("Statement should contain one and only one statement"));
  }

  if (list_length(cte_stmts) != 1) {
    ereport(ERROR, errmsg("CTE should contain one and only one statement"));
  }

  void *node = linitial(stmts);

  CommonTableExpr *cte_node = palloc0(sizeof(*cte_node));

  *cte_node = (CommonTableExpr) {
    .type = T_CommonTableExpr, .ctename = text_to_cstring(cte_name), .aliascolnames = NULL,
    .ctematerialized = CTEMaterializeDefault, .ctequery = linitial_node(RawStmt, cte_stmts)->stmt,
#if PG_VERSION_MAJORNUM >= 14
    .search_clause = NULL, .cycle_clause = NULL,
#endif
    .location = -1, // unknown location
        .cterecursive = recursive, .cterefcount = 0, .ctecolnames = NULL, .ctecoltypes = NULL,
    .ctecoltypmods = NULL, .ctecolcollations = NULL
  };

  WithClause **with = NULL;
dispatch:
  switch (nodeTag(node)) {
  case T_RawStmt: {
    RawStmt *stmt = castNode(RawStmt, node);
    node = stmt->stmt;
    goto dispatch;
  }
  case T_SelectStmt: {
    SelectStmt *select = castNode(SelectStmt, node);
    with = &select->withClause;
    break;
  }
  case T_InsertStmt: {
    InsertStmt *insert = castNode(InsertStmt, node);
    with = &insert->withClause;
    break;
  }
  case T_UpdateStmt: {
    UpdateStmt *update = castNode(UpdateStmt, node);
    with = &update->withClause;
    break;
  }
  case T_DeleteStmt: {
    DeleteStmt *delete = castNode(DeleteStmt, node);
    with = &delete->withClause;
    break;
  }
  default:
    ereport(ERROR, errmsg("no supported statement found"));
  }

  if (*with == NULL) {
    WithClause *new_with = palloc(sizeof(*new_with));
    new_with->type = T_WithClause;
    new_with->location = -1;
    new_with->recursive = recursive;
    new_with->ctes = list_make1(cte_node);
    *with = new_with;
  } else {
    List *ctes = (*with)->ctes;
    if (prepend) {
      ctes = list_insert_nth(ctes, 0, cte_node);
    } else {
      ctes = lappend(ctes, cte_node);
    }
  }

  return stmts;
}

static bool contains_param_walker(Node *node, bool *contains) {
  if (node != NULL) {
    if (nodeTag(node) == T_ParamRef) {
      *contains = true;
      return true;
    } else {
      return raw_expression_tree_walker(node, contains_param_walker, contains);
    }
  }
  return false;
}

bool omni_sql_is_parameterized(List *stmts) {

  ListCell *stmt;
  foreach (stmt, stmts) {
    bool contains = false;
    raw_expression_tree_walker(castNode(RawStmt, lfirst(stmt))->stmt, contains_param_walker,
                               &contains);
    if (contains) {
      return true;
    }
  }
  return false;
}