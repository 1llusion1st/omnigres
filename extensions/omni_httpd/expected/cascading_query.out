CREATE TABLE routes (
   name text NOT NULL,
   query text NOT NULL,
   priority INT NOT NULL
);
INSERT INTO routes (name, query, priority) VALUES
  ('test', $$SELECT omni_httpd.http_response(body => 'test') FROM request WHERE request.path = '/test'$$, 1);
INSERT INTO routes (name, query, priority) VALUES
  ('ping', $$SELECT omni_httpd.http_response(body => 'pong') FROM request WHERE request.path = '/ping'$$, 1);
\pset format wrapped
\pset columns 80
-- Preview the query
SELECT omni_httpd.cascading_query(name, query) FROM routes GROUP BY priority ORDER BY priority DESC;
                                cascading_query                                 
--------------------------------------------------------------------------------
 WITH test AS (SELECT omni_httpd.http_response(body := 'test') FROM request WHE.
.RE request.path = '/test'), ping AS (SELECT omni_httpd.http_response(body := '.
.pong') FROM request WHERE request.path = '/ping') SELECT * FROM test UNION ALL.
. SELECT * FROM ping WHERE NOT EXISTS (SELECT FROM test)
(1 row)

\pset format aligned
-- Try it
BEGIN;
WITH listener AS (INSERT INTO omni_httpd.listeners (address, port) VALUES ('127.0.0.1', 9100) RETURNING id),
     handler AS (INSERT INTO omni_httpd.handlers (query) SELECT omni_httpd.cascading_query(name, query) FROM routes GROUP BY priority ORDER BY priority DESC RETURNING id)
INSERT INTO omni_httpd.listeners_handlers (listener_id, handler_id)
SELECT listener.id, handler.id
FROM listener, handler;
DELETE FROM omni_httpd.configuration_reloads;
END;
CALL omni_httpd.wait_for_configuration_reloads(1);
\! curl --retry-connrefused --retry 10  --retry-max-time 10 -w '\n\n' --silent http://localhost:9100/test
test

\! curl --retry-connrefused --retry 10  --retry-max-time 10 -w '\n\n' --silent http://localhost:9100/ping
pong

